### PR TITLE
Hide external buffer types used by `primitive::Encoder` in aws-smithy-types

### PR DESCRIPTION
**Pull request by jdisanti**
## Motivation and Context
The buffer types are an implementation detail for the primitive encoding, and should not be tied to the public API. This PR wraps them in some lightweight structs to hide them.

## Testing
- Added `api-linter` to CI for the `aws-smithy-types` crate now that it reveals no external types in its API

## Checklist
- [x] I have updated `CHANGELOG.next.toml` if I made changes to the smithy-rs codegen or runtime crates

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: rust-runtime/aws-smithy-types/src/primitive.rs [hunk 1]
```diff
@@ -107,25 +119,25 @@ pub enum Encoder {
     Bool(bool),
     /// 8-bit signed integer
     #[non_exhaustive]
-    I8(i8, itoa::Buffer),
```
Comment (rcoh) [inline rust-runtime/aws-smithy-types/src/primitive.rs:110]: do we actually even need these public variants on encoder? Can we do `struct Encoder(Inner); `enum Inner { ... }`?

Comment (rcoh) [reply rust-runtime/aws-smithy-types/src/primitive.rs:110]: since we have the from impls already

**Review by rcoh — APPROVED** [review]
lgtm, lmk what you think about just making it a private tuple struct though

File: rust-runtime/aws-smithy-types/src/primitive.rs (hunk 1 above)
Comment (jdisanti) [reply rust-runtime/aws-smithy-types/src/primitive.rs:110]: I like this idea! Changed over.

**Review by rcoh — APPROVED** [review]

**Pull request merged**